### PR TITLE
Initial Setup

### PR DESCRIPTION
**Pull request by wurli**
This PR re-adds previous work which was pushed directly to `main` so that it can be properly reviewed.

* Adds new functions `dfe_cols()`, `dfe_pal()`, `scale_colour_dfe()` and `scale_fill_dfe()`
* Adds a brief `README.md`

### CODE REVIEW

**Comment by wurli** [comment]
Before merging `devtools::check()` will need to complete successfully - currently it produces the following:

```
── R CMD check results ────────────────────────────────────────────────────────────────────────────────────────────────────────────── dfethemes 0.0.0.9000 ────
Duration: 35.7s

❯ checking examples ... ERROR
  Running examples in 'dfethemes-Ex.R' failed
  The error most likely occurred in:
  
  > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
  > ### Name: scale_fill_dfe
  > ### Title: Add fill scale to ggplot with DfE theme
  > ### Aliases: scale_fill_dfe
  > 
  > ### ** Examples
  > 
  > library(ggplot2)
  > iris |>
  +  ggplot(aes(x=Species,y=Sepal.Length,fill=Species))+
  +  geom_col()+
  +  scale_fill_dfe()
  > 
  > 
  > data <- data.frame(murder = USArrests$Murder,
  +  state = tolower(rownames(USArrests)))
  >  map <- map_data("state")
  Error in map_data("state") : 
    The package `maps` is required for `map_data()`
  Calls: map_data -> check_installed
  Execution halted

❯ checking top-level files ... NOTE
  Non-standard file/directory found at top level:
    'assets'

1 error ✖ | 0 warnings ✔ | 1 note ✖
```

File: README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# dfethemes
+
+This package seeks to create an easy theme to add to any ggplot to keep it in line with ['DfE brand guidelines'](https://educationgovuk.sharepoint.com/sites/how-do-i/SitePages/communications-how-to-use-branding-in-the-department-and-its-executive-agencies.aspx?xsdata=MDV8MDF8fGFkNzQzOGRkZWUwNjQ5NWRlOTIwMDhkYWJjMmE0ZTQ1fGZhZDI3N2M5YzYwYTRkYTFiNWYzYjNiOGIzNGE4MmY5fDF8MHw2MzgwMjkxODcxMTQxOTkzMjh8R29vZHxWR1ZoYlhOVFpXTjFjbWwwZVZObGNuWnBZMlY4ZXlKV0lqb2lNQzR3TGpBd01EQWlMQ0pRSWpvaVYybHVNeklpTENKQlRpSTZJazkwYUdWeUlpd2lWMVFpT2pFeGZRPT18MXxNVGs2TXpFeFpXTXlaVFJrTkRaaU5HUmtNemhtTUdRMk1XWXdOV1ppT1RNek9ETkFkR2h5WldGa0xuTnJlWEJsfHw%3D&sdata=YzJ6TWJBT0Z1cXduTUhsenZIQ1E5VzUrSjM2Q3lDZlN1TWRvZlNvdWhYMD0%3D#using-the-brand).
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply README.md]: Removed this

File: README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+warm - Red to purple (continuous)
+cool - Blue to lime (continuous)
+full - all colours in the brand (discrete, 7 colours)
+likert - red to blue, grey midpoint (diverging)
+likert2 - green to purple, white midpoint (diverging)
+likert3 -  red to blue, white midpoint (diverging)
+heat - white to red (continuous)
+heat2 - white to pink (continuous)
+cold - white to blue (continuous)
+cold2 - white to turqouise (continuous)
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply README.md]: Added markdown bullets

File: README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+This package seeks to create an easy theme to add to any ggplot to keep it in line with ['DfE brand guidelines'](https://educationgovuk.sharepoint.com/sites/how-do-i/SitePages/communications-how-to-use-branding-in-the-department-and-its-executive-agencies.aspx?xsdata=MDV8MDF8fGFkNzQzOGRkZWUwNjQ5NWRlOTIwMDhkYWJjMmE0ZTQ1fGZhZDI3N2M5YzYwYTRkYTFiNWYzYjNiOGIzNGE4MmY5fDF8MHw2MzgwMjkxODcxMTQxOTkzMjh8R29vZHxWR1ZoYlhOVFpXTjFjbWwwZVZObGNuWnBZMlY4ZXlKV0lqb2lNQzR3TGpBd01EQWlMQ0pRSWpvaVYybHVNeklpTENKQlRpSTZJazkwYUdWeUlpd2lWMVFpT2pFeGZRPT18MXxNVGs2TXpFeFpXTXlaVFJrTkRaaU5HUmtNemhtTUdRMk1XWXdOV1ppT1RNek9ETkFkR2h5WldGa0xuTnJlWEJsfHw%3D&sdata=YzJ6TWJBT0Z1cXduTUhsenZIQ1E5VzUrSjM2Q3lDZlN1TWRvZlNvdWhYMD0%3D#using-the-brand).
+
+## Colours
+
+There are two functions that work to add in colour scales: `scale_fill_dfe()` and `scale_colour_dfe()`. These replace the usual `scale_colour_*()` and `scale_fill_*()` functions to change colours on ggplot. 
+
+These are the base colours
+
+<img src="/assets/dfe_colour_pallete.png">
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply README.md]: Created a function `display.dfe.all()` which fulfils this. Also useful for displaying the palettes, not dissimilar to `RColorBrewer::display.brewer.all()`.

File: R/dfe_colours.R [hunk 4]
```diff
@@ -0,0 +1,148 @@
+require(tidyverse)
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply R/dfe_colours.R]: Removed this and added them in description/namespace. I have used some others in the `display.dfe.all()` function so added those in.

File: R/dfe_colours.R [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                             ...){
+  pal <- dfe_pal(palette = palette, reverse = reverse)
+
+    if (discrete){
+      ggplot2::discrete_scale("colour", paste0("dfe_", palette), palette = pal, ...)
+    } else {
+      ggplot2::scale_colour_gradientn(colours = pal(256))
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply R/dfe_colours.R:104]: On these:

- added in a check that the palettes is in `dfe_palettes` and returns an 'unrecognised palette' message if it's an invalid selection.
- I've done some searching and I can't find a good explanation on why there are two versions for discrete/continuous versions for the ggplot universe colour selection - however like you say it's probably best to copy this. I've not done it yet however as maybe it's worth a conversation about if we want to have ones for the three main colour uses, ie, continuous 1 colour, diverging/three-colour continuous, and categorial/qualiatative?
- Added in markdown bullets

File: R/dfe_colours.R [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+#'
+scale_fill_dfe <- function(palette = "main", discrete = TRUE, reverse = FALSE, ...) {
+  pal <- dfe_pal(palette = palette, reverse = reverse)
+
+  if (discrete) {
+    ggplot2::discrete_scale("fill", paste0("dfe_", palette), palette = pal, ...)
+  } else {
+    ggplot2::scale_fill_gradientn(colours = pal(256), ...)
+  }
+}
```
[the remark this answers is not in the document]
Comment (scookedfe) [reply R/dfe_colours.R]: Resolved

**Review by scookedfe — COMMENTED** [review]
Made changes - let me know what you think

**Comment by wurli** [comment]
Hey @scookedfe, all looking good. I've created a PR into this branch with a bunch of minor edits - please take a look and feel free to merge it if you're happy. 

Once you've looked at that, there are a few slightly more substantial changes which I think would be worthwhile:

* Your code makes use of 3 fairly big package packages, tibble, dplyr and purrr. I think it would be worthwhile to remove these dependencies if possible (looking at your code I don't think this would be too hard to do). The reason for this is that it's good practice to make packages as lightweight as possible - if someone wants to use ggdfe, we'd want them to be able to install and load it as fast as possible, and we wouldn't want ggdfe to suddenly break due to a change to another package if it can be avoided. Balancing dependencies is a [tricky issue](https://www.tidyverse.org/blog/2019/05/itdepends/), but I think in this case they can probably be removed.
* There's a fairly chunky bit of documentation which is reused in your code - the `@return` value for the colour scales. Roxygen2 has some [nice functionality](https://roxygen2.r-lib.org/articles/reuse.html) for reusing documentation, so you might want to take advantage of that here in case these colours change in the future - updating the docs in multiple places will be time consuming and prone to mistakes. I wouldn't say this is high priority because currently there's only one bit of duplication, but it's something to be aware of.
* In `scale_colour_dfe()` you use `cat()` to throw a message when the `palette` arg isn't valid - to me this feels like it should throw an error instead. I'd recommend using the *cli* package to do this:
  * I'd import the whole of cli because its functions are so useful. You can do this by adding a line like `#' @import cli` in `ggdfe-package.R`
  * The actual function I'd then use to throw an error would be `cli_abort()` (the base equivalent would be `stop()`. `cli_abort()` has really handy formatting/interpolation abilities, so you can do stuff like this (try in RStudio to see nicer formatting):
    ``` r
    library(cli)
    
    valid <- c("a", "b", "c", "d", "e")
    
    f <- function(x) {
      if (!x %in% valid) {
        cli_abort(c(
          "{.val {x}} is not a valid palette",
          i = "Valid palettes are {.val {valid}}"
        ))
      } 
    }
    
    f("brian")
    #> Error in `f()`:
    #> ! "brian" is not a valid palette
    #> ℹ Valid palettes are "a", "b", "c", "d", and "e"
    
    #> Backtrace:
    #>     ▆
    #>  1. └─global f("brian")
    #>  2.   └─cli::cli_abort(c("{.val {x}} is not a valid palette", i = "Valid palettes are {.val {valid}}"))
    #>  3.     └─rlang::abort(...)
    ```

    <sup>Created on 2022-11-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
* As functions become more complex it'd be good to add some [unit testing](https://r-pkgs.org/testing-basics.html). I'd recommend adding some basic tests for the `scale_` functions at least, as they're user-facing. Perhaps tests to check that they fail correctly and produce objects of the appropriate type correctly would be enough, as they're fairly simple functions.

**Comment by scookedfe** [comment]
Hi @wurli, I've merged the initial-setup edits and made some changes.

- simplified `show_dfe_palettes` so that it doesn't use other packages from the tidyverse.
- used `@describeIn` for scale_fill_dfe to combine. 
- used your `cli_abort` code as a function to check if palettes are valid. I've added in an argument that defaults to dfe_palettes as the set of palettes to check against, thinking we can reuse this further down the line if we want to have a set of palettes associated with dfe or .gov styles.

I'll have a look at unit testing in a short while.

**Comment by wurli** [comment]
LGTM 👍

**Pull request merged**